### PR TITLE
Pretty printing dataset

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Description
Implementation of one-line approach described here: https://github.com/kedro-org/kedro/issues/3980

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/io/core.py [hunk 1]
```diff
@@ -227,32 +228,18 @@ def save(self, data: _DI) -> None:
...
+            if object_description[arg_name] is not None:
+                str_keys.append(
+                    f"{arg_name}={pprint.pformat(object_description[arg_name], compact=True)}"
                 )
 
-                return text if is_root else "{" + text + "}"  # 1
-
-            # not a dictionary
-            return str(obj)
+        return f"{type(self).__module__}.{type(self).__name__}({', '.join(str_keys)})"
```
Comment (deepyaman) [inline kedro/io/core.py:239]: I think the issue with this is that all of the pretty-printed subkeys will be on a single line, and no formatting/wrapping is applied there. By using `PrettyPrinter` directly, you can get something pretty similar, and also get the formatting.

Happy to share some more context/pointers on this tomorrow or next week, if you'd like.

Comment (ElenaKhaustova) [reply kedro/io/core.py:239]: I played a bit with `PrettyPrinter` and different indentations but in the end, I preferred one line to it. There shouldn't be too many keys as they are some subset of class constructor input arguments passed via `_describe()`. I was planning to have this as a bare minimum but let me add a second option with the indentation, so we can choose which one looks better.

Happy to discuss them tomorrow.

File: kedro/io/core.py [hunk 2]
```diff
@@ -227,32 +228,23 @@ def save(self, data: _DI) -> None:
...
+                    depth=2,
+                    width=500,
                 )
+                str_keys.append(f"{arg_name}={descr}")
 
-                return text if is_root else "{" + text + "}"  # 1
-
-            # not a dictionary
-            return str(obj)
+        return f"{type(self).__module__}.{type(self).__name__}({', '.join(str_keys)})"
```
Comment (deepyaman) [inline kedro/io/core.py:244]: If it comes from `kedro_datasets`, I feel like it would be much cleaner to have the short name users can provide (i.e. `pandas.CSVDataset` instead of `kedro_datasets.pandas.csv_dataset.CSVDataset`).

Similarly, from Kedro core, `MemoryDataset`, etc. seems much easier to read than the full path.

File: kedro/io/core.py [hunk 3]
```diff
@@ -227,32 +228,23 @@ def save(self, data: _DI) -> None:
...
-            if isinstance(obj, dict):
-                sorted_dict = sorted(obj.items(), key=lambda pair: str(pair[0]))  # 2
-
-                text = ", ".join(
-                    fmt.format(key, _to_str(value))  # 2
-                    for key, value in sorted_dict
-                    if value is not None  # 3
+    def _pretty_repr(self, object_description: dict[str, Any]) -> str:
+        str_keys = []
+        for arg_name in sorted(object_description, key=lambda key: str(key)):
```
Comment (deepyaman) [inline kedro/io/core.py:233]: They key is already a string, right?

Comment (deepyaman) [reply kedro/io/core.py:233]:
```suggestion
        for key, value in sorted(object_description.items()):
```
feels a lot more Pythonic

Nit: `obj` is a perfectly Pythonic name for these cases, and easier than `object_description` IMO

File: kedro/io/core.py [hunk 4]
```diff
@@ -227,32 +228,23 @@ def save(self, data: _DI) -> None:
...
+            if object_description[arg_name] is not None:
+                descr = pprint.pformat(
+                    object_description[arg_name],
+                    sort_dicts=False,
+                    compact=True,
+                    depth=2,
+                    width=500,
```
Comment (deepyaman) [inline kedro/io/core.py:240]:
```suggestion
                    width=float("inf"),
```
rather than random value?

File: kedro/io/core.py [hunk 5]
```diff
@@ -227,32 +228,23 @@ def save(self, data: _DI) -> None:
...
-                    fmt.format(key, _to_str(value))  # 2
-                    for key, value in sorted_dict
-                    if value is not None  # 3
+    def _pretty_repr(self, object_description: dict[str, Any]) -> str:
+        str_keys = []
+        for arg_name in sorted(object_description, key=lambda key: str(key)):
+            if object_description[arg_name] is not None:
+                descr = pprint.pformat(
+                    object_description[arg_name],
+                    sort_dicts=False,
```
Comment (deepyaman) [inline kedro/io/core.py:237]:
```suggestion
```
Although, why sort the top-level key, but not subkeys?

**Review by deepyaman — CHANGES_REQUESTED** [review]
What does the pretty repr for a `PartitionedDataset` look like? E.g. from the docs:

```yaml
# conf/base/catalog.yml

my_partitioned_dataset:
  type: partitions.PartitionedDataset
  path: s3://my-bucket-name/path/to/folder
  dataset:  # full dataset config notation
    type: pandas.CSVDataset
    load_args:
      delimiter: ","
    save_args:
      index: false
  credentials: my_credentials
  load_args:
    load_arg1: value1
    load_arg2: value2
  filepath_arg: filepath  # the argument of the dataset to pass the filepath to
  filename_suffix: ".csv"
```

**Comment by datajoely** [comment]
Can you include an example of what it looks like?

File: kedro/io/core.py (hunk 5 above)
Comment (ElenaKhaustova) [reply kedro/io/core.py:237]: I wanted to keep them in the order provided in the config, so removing top-level key sorting for consistency.

**Comment by ElenaKhaustova** [comment]
> ```yaml
> my_partitioned_dataset:
>   type: partitions.PartitionedDataset
>   path: s3://my-bucket-name/path/to/folder
>   dataset:  # full dataset config notation
>     type: pandas.CSVDataset
>     load_args:
>       delimiter: ","
>     save_args:
>       index: false
>   credentials: my_credentials
>   load_args:
>     load_arg1: value1
>     load_arg2: value2
>   filepath_arg: filepath  # the argument of the dataset to pass the filepath to
>   filename_suffix: ".csv"
> ```

Here is how `_describe()` is implemented, so credentials are omitted: https://github.com/kedro-org/kedro-plugins/blob/be99fecf6cf5ac8f6a0a717c56b06dbc148b26eb/kedro-datasets/kedro_datasets/partitions/partitioned_dataset.py#L318

Screenshot 2024-07-09 at 12 23 18

**Comment by ElenaKhaustova** [comment]
> Can you include an example of what it looks like?

We used one-line approach described here: https://github.com/kedro-org/kedro/issues/3980#issuecomment-2209639882

Here is the updated representation after removing sorting:
Screenshot 2024-07-09 at 12 30 27

File: kedro/io/core.py (hunk 2 above)
Comment (ElenaKhaustova) [reply kedro/io/core.py:245]: I agree on your points but decided to keep the full path, so `__repr__` is as ambiguous as possible and one could easily understand where to look the implementation of the printed dataset.

Ideally, it would be good to have some `"short=True"` flag for the representation that can be set by user. But by default, I prefer to keep a full module name.

Curious what other think.

File: kedro/io/core.py (hunk 1 above)
Comment (ElenaKhaustova) [reply kedro/io/core.py:245]: Two considered options: https://github.com/kedro-org/kedro/issues/3980

File: kedro/io/core.py [hunk 6]
```diff
@@ -227,32 +229,23 @@ def save(self, data: _DI) -> None:
             message = f"Failed while saving data to data set {str(self)}.\n{str(exc)}"
             raise DatasetError(message) from exc
 
-    def __str__(self) -> str:
```
Comment (astrojuanlu) [inline kedro/io/core.py:230]: IIUC, we're removing the `__str__` format right? Shouldn't we keep it around for backwards compatibility?

File: kedro/io/core.py (hunk 2 above)
Comment (deepyaman) [reply kedro/io/core.py:245]: Happy to hear what others think, but IMO we actually don't use the module name in the docs and examples. The standard way people write entries in catalog is the "short" version,  and I think that should be reflected.

**Comment by deepyaman** [comment]
> > Can you include an example of what it looks like?
> 
> We used one-line approach described here: [#3980 (comment)](https://github.com/kedro-org/kedro/issues/3980#issuecomment-2209639882)
> 
> Here is the updated representation after removing sorting: Screenshot 2024-07-09 at 12 30 27

Would make more sense for `PartitionedDataset` to have a `dataset` key that is formatted like a dataset repr, for consistency with `CachedDataset`.

For `CachedDataset`, the wrapped dataset represents should also not be in quotes I think.

File: kedro/io/core.py (hunk 6 above)
Comment (ElenaKhaustova) [reply kedro/io/core.py:230]: If `__str__` is not defined the default is to use the `__repr__`, so it shouldn't be a problem.

File: kedro/io/core.py (hunk 2 above)
Comment (ElenaKhaustova) [reply kedro/io/core.py:245]: We can also consider adding `__str__` with a short representation, so that `__repr__` is actual representation but `__str__` adapted for printing

**Comment by ElenaKhaustova** [comment]
> Would make more sense for `PartitionedDataset` to have a `dataset` key that is formatted like a dataset repr, for consistency with `CachedDataset`.

Happy to do that later on.

**Comment by ElenaKhaustova** [comment]
> For `CachedDataset`, the wrapped dataset represents should also not be in quotes I think.

```
    def _describe(self) -> dict[str, Any]:
        return {
            "dataset": self._dataset._pretty_repr(self._dataset._describe()),
            "cache": self._cache._pretty_repr(self._cache._describe()),
        }
```

The idea was to keep `"dataset"` and `"cache"` representations aligned with dataset representation - `Class(arg_1=val_1, ...)` , so the values for `"dataset"` and `"cache"` keys are converted to the corresponding strings. That's why I cannot easily get rid of quotes.

File: kedro/io/core.py (hunk 6 above)
Comment (astrojuanlu) [reply kedro/io/core.py:230]: Oh right. Still, we're changing __str__ right? Not sure if someone depends on str(dataset). it looks like a small, preventable breaking change.

Comment (ElenaKhaustova) [reply kedro/io/core.py:230]: So you mean keeping the old `__str__`?

Comment (astrojuanlu) [reply kedro/io/core.py:230]: Yes, at least until 0.20...

**Comment by deepyaman** [comment]
@astrojuanlu @merelcht @noklam @datajoely would love to get your view on https://github.com/kedro-org/kedro/pull/3987#discussion_r1669430539 especially. 

Historically, I think the direct module name has been an implementation detail, and not something we surface to users (e.g. in the docs). If anything, it's the sort of thing you might stick under an underscore prefix, were we to adopt thay convention.

My personal preference would be for this aspect to remain closer to the shorter representation we have now, but would be great to get some thoughts. Think it would be good to not switch it and then change it back in another release.

**Comment by datajoely** [comment]
I guess the argument in favour is so that someone can easily import class in a notebook and test things.

I sort of outline this in this prototype https://github.com/kedro-org/kedro/issues/3980#issuecomment-2217989257

**Comment by noklam** [comment]
I am not a big fan of the shortcut, as this create burdens that Kedro need to be smart about guessing the import only to save some shortcut. This has more value in the past with a few different modules of dataset, but as most dataset moved to `kedro-datasets`, the main value is keep it non-breaking IMO.

I think, with the de-coupling of framework/library, it is better to not assume datasets is being used with framework. There are also side benefit of not trying to hide the import as @datajoely described.

TL;DR The standalone usage of dataset is the most important one to me and I slightly lean to the current implementation.

File: kedro/io/core.py [hunk 7]
```diff
@@ -234,6 +236,8 @@ def _to_str(obj: Any, is_root: bool = False) -> str:
             formatted like Dataset(key=value).
             2. Dictionaries have the keys alphabetically sorted recursively.
             3. None values are not shown.
+
+            Will be removed in 0.20.0 release.
```
Comment (merelcht) [inline kedro/io/core.py:240]: Is this a deprecation meant to be seen by users? It should at least be mentioned in the release notes, but usually we throw a deprecation warning as well, see: https://github.com/kedro-org/kedro/blob/main/kedro/framework/cli/utils.py#L198-L200

**Comment by merelcht** [comment]
I fully agree with @noklam and also prefer the longer represenation.

**Review by merelcht — APPROVED** [review]
I left one comment about the deprecation message, but otherwise looks great! 👍

File: kedro/io/core.py (hunk 7 above)
Comment (ElenaKhaustova) [reply kedro/io/core.py]: This comment refers to a private method, so it's not for users to see but for us. I replaced it with TODO for clarity. The aim is to remove the current `__str__` implementation, so `__repr__` is used by default, so it's not a deprecation but an update.

File: kedro/io/core.py (hunk 2 above)
Comment (astrojuanlu) [reply kedro/io/core.py:273]: > The standard way people write entries in catalog is the "short" version

Most users we've spoken to create their own datasets sooner or later, so people have a mix between short and long names in the catalog.

Also, the long name would work even for `kedro-datasets` right? The fact that we use the short one is a matter of convention and convenience. But arguably it obscures where the datasets come from, and maybe play a role in people not understanding their tracebacks (especially with `s` vs `S`).

So I'm slightly in favour of keeping the long names for all datasets to avoid introducing an exception to the logic written here, but I don't have a strong opinion.

(In either case, no matter what we do I don't think we should complicate the implementation with a flag.)

**Review by astrojuanlu — APPROVED** [review]
Approved, contingent on resolving the discussion on long vs short-form 👍🏼

**Comment by deepyaman** [comment]
> > Would make more sense for `PartitionedDataset` to have a `dataset` key that is formatted like a dataset repr, for consistency with `CachedDataset`.
> 
> Happy to do that later on.

Can you create an issue for this, if not already done?

**Comment by ElenaKhaustova** [comment]
> > > Would make more sense for `PartitionedDataset` to have a `dataset` key that is formatted like a dataset repr, for consistency with `CachedDataset`.
> > 
> > 
> > Happy to do that later on.
> 
> Can you create an issue for this, if not already done?

Issue created: https://github.com/kedro-org/kedro-plugins/issues/769
